packages/dev/core/src/XR/features/WebXRSpaceWarp.ts

```diff
@@ -0,0 +1,367 @@
+                }
+            );
+            this._velocityMaterial._materialHelperNeedsPreviousMatrices = true;
+            this._velocityMaterial.onBindObservable.add((mesh) => {
+                // mesh. getWorldMatrix can be incorrect under rare conditions (e.g. when using a effective mesh in the render function).
+                // If the case arise that will require changing it we will need to change the bind process in the material class to also provide the world matrix as a parameter
+                this._previousWorldMatrices[mesh.uniqueId] = this._previousWorldMatrices[mesh.uniqueId] || mesh.getWorldMatrix();
+                this._velocityMaterial.getEffect().setMatrix("previousWorld", this._previousWorldMatrices[mesh.uniqueId]);
+                this._previousWorldMatrices[mesh.uniqueId] = mesh.getWorldMatrix();
+                // now set the scene's previous matrix
+                this._velocityMaterial.getEffect().setMatrix("previousViewProjection", this._previousTransforms[0]);
+                // multiview for sure
+                this._velocityMaterial.getEffect().setMatrix("previousViewProjectionR", this._previousTransforms[1]);
+
+                // store the previous (current, to be exact) transforms
+                this._previousTransforms[0].copyFrom(scene.getTransformMatrix());
+                this._previousTransforms[1].copyFrom(scene._transformMatrixR);
+            });
+            this._velocityMaterial.freeze();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public render(useCameraPostProcess: boolean = false, dumpForDebug: boolean = false): void {
+        // Swap to use velocity material
+        this._originalPairing.length = 0;
+        const scene = this.getScene();
+        // set the velocity material to render the velocity RTT
+        if (scene && this._velocityMaterial) {
+            scene.getActiveMeshes().forEach((mesh) => {
+                this._originalPairing.push([mesh, mesh.material]);
+                mesh.material = this._velocityMaterial;
+            });
+        }
+
+        super.render(useCameraPostProcess, dumpForDebug);
+
+        // Restore original materials
+        this._originalPairing.forEach((tuple) => {
+            tuple[0].material = tuple[1];
+        });
+    }
+
+    /**
+     * @internal
+     */
+    public _bindFrameBuffer() {
+        if (!this._renderTarget) {
+            return;
+        }
+        this.getScene()!.getEngine().bindSpaceWarpFramebuffer(this._renderTarget);
+    }
+
+    /**
+     * Gets the number of views the corresponding to the texture (eg. a SpaceWarpRenderTarget will have > 1)
+     * @returns the view count
+     */
+    public getViewCount() {
+        return 2;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public dispose(): void {
+        super.dispose();
+        this._velocityMaterial.dispose();
+        this._previousTransforms.length = 0;
+        this._previousWorldMatrices.length = 0;

```

Comment:
Should we clear _originalPairing here too? I see it's cleared in render already, but in case dispose is called between renders it could not be cleared?

Reply:
Done :-)